File: CallbackListener/app/src/main/java/com/example/callbacklistener/MainActivity.java
package com.example.callbacklistener;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        listener.onEvent();
    }

    MyEventListener listener = new MyEventListener() { // 인터페이스는 한번 호출하면 코드가 모두 전시. 그리고 변경이 불가능하다.
        @Override
        public void onEvent() {

        }
    };

//    MyEventClass listener2 = listener.onEvent(); 클래스 호출은 객체를 따로 불러올 수 있다.

//    MyEventClass listener3 = new MyEventClass().onEvent2(); 클래스 호출은 객체를 따로 불러올 수 있다.


}

File: AACMVVM/app/src/main/java/com/puresoftware/aacmvvm/MainActivity.kt
package com.puresoftware.aacmvvm

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.puresoftware.aacmvvm.databinding.ActivityMainBinding

// Activity
class MainActivity : AppCompatActivity() {

    lateinit var binding: ActivityMainBinding
    lateinit var viewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        // instance
        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        binding.viewModel = viewModel

        // ViewModel에 database 넣기
        viewModel.setDatabase(ModelDatabase())

        // 람다식 기반의
        viewModel.info.observe(this) {
            binding.viewModel = viewModel
        }

        //        // observe. 구버전
//        viewModel.info.observe(this, Observer {
//            binding.text.text = it // 이건 viewBinding으로 쓰면 조건이 맞지 않다.
//        })
    }
}